en, zh: add and update tiup-dm docs (#264)
pingcap/docs-dm: en/maintain-dm-using-tiup.md

---
title: Maintain a DM Cluster Using TiUP
summary: Learn how to maintain a DM cluster using TiUP.
---

# Maintain a DM Cluster Using TiUP

This document introduces how to maintain a DM cluster using the TiUP DM component. For the complete steps of DM deployment, refer to [Deploy a DM Cluster Using TiUP](deploy-a-dm-cluster-using-tiup.md).

For the help information of the TiUP DM component, run the following command:

```bash
tiup dm --help
```

```
Deploy a DM cluster for production

Usage:
  tiup dm [flags]
  tiup dm [command]

Available Commands:

>   deploy      Deploy a DM cluster for production

Author: @csuzhangxc Is this section directly from the terminal?

172.19.0.101:3000  grafana       172.19.0.101  3000       linux/x86_64  Up         -                                  /home/tidb/deploy/grafana-3000
172.19.0.101:9090  prometheus    172.19.0.101  9090       linux/x86_64  Up         /home/tidb/data/prometheus-9090    /home/tidb/deploy/prometheus-9090
```

The `Status` column uses `Up` or `Down` to indicate whether the service is running normally.

For the DM-master component, `|L` might be appended to a status, which indicates that the DM-master node is a Leader. For the DM-worker component, `Free` indicates that the current DM-worker node is not bound to an upstream.

## Scale in a node

Scaling in a node means taking the node offline. This operation removes the node from the cluster and deletes the remaining data files.

When you scale in a node, DM operations on DM-master and DM-worker components are performed in the following order:

1. Stop component processes.
2. Call the API for DM-master to delete the `member`.
3. Clean up the data files related to the node.

The basic usage of the scale-in command:

```bash
tiup dm scale-in <cluster-name> -N <node-id>
```

> To use this command, you need to specify at least two flags: the cluster name and the node ID. The node ID can be obtained by using the `tiup dm display` command in the previous section.

Reviewer: ```suggestion
To use this command, you need to specify at least two arguments: the cluster name and the node ID. The node ID can be obtained by using the `tiup dm display` command in the previous section.
```

For example, to scale out a DM-worker node in the `prod-cluster` cluster, take the following steps:

1. Create a `scale.yaml` file and add information of the new worker node:

    > **Note:**
    >
    > You need to create a topology file, which includes only the description of the new nodes, not the existing nodes.

    ```yaml
    ---

    worker_servers:
      - host: 172.16.5.140

    ```

2. Perform the scale-out operation. TiUP DM adds the corresponding nodes to the cluster according to the port, directory, and other information described in `scale.yaml`.

    {{< copyable "shell-regular" >}}

    ```shell
    tiup dm scale-out prod-cluster scale.yaml
    ```

    After the command is executed, you can check the status of the scaled-out cluster by running `tiup dm display prod-cluster`.

## Rolling upgrade

The rolling upgrade process is made as transparent as possible to the application, and does not affect the business. The operations vary with different nodes.

### Upgrade command

> For example, the following command upgrades the cluster to v2.0.1:

Author: ```suggestion
You can run the `tiup dm upgrade` command to upgrade a DM cluster. For example, the following command upgrades the cluster to v2.0.1:
```

      --wait-timeout int   Timeout in seconds to wait for an operation to complete, ignored for operations that don't fit. (default 60)
  -y, --yes                Skip all confirmations and assumes 'yes'
```

If a DM-master hotfix package is in `/tmp/dm-master-hotfix.tar.gz` and you want to replace all the DM-master packages in the cluster, run the following command:

{{< copyable "shell-regular" >}}

```bash
tiup dm patch prod-cluster /tmp/dm-master-hotfix.tar.gz -R dm-master
```

You can also replace only one DM-master package in the cluster:

{{< copyable "shell-regular" >}}

```bash
tiup dm patch prod-cluster /tmp/dm--hotfix.tar.gz -N 172.16.4.5:8261
```

## Import and upgrade a DM 1.0 cluster deployed using DM-Ansible

Before TiUP is released, DM-Ansible is often used to deploy DM clusters. To enable TiUP to take over the DM 1.0 cluster deployed by DM-Ansible, use the `import` command.

For example, to import a cluster deployed using DM Ansible:

{{< copyable "shell-regular" >}}

```bash
tiup dm import --dir=/path/to/tidb-ansible
```

The process of using the `import` command is as follows:

> 1. TiUP generates a topology file [`topology.yml`](https://github.com/pingcap/tiup/blob/master/examples/topology.dm.example.yaml) based on the DM cluster previously deployed using DM Ansible.

Author: ```suggestion
1. TiUP generates a topology file [`topology.yml`](https://github.com/pingcap/tiup/blob/master/examples/topology.dm.example.yaml) based on the DM cluster previously deployed using DM-Ansible.
```

2. After confirming that the topology file has been generated, you can use it to deploy the DM cluster of v2.0 or later versions.

After the deployment is completed, you can execute the `tiup dm start` command to start the cluster and begin the process of upgrading the DM kernel.

## View the operation log

To view the operation log, use the `audit` command. The usage of the `audit` command is as follows:

```bash
Usage:
  tiup dm audit [audit-id] [flags]

Flags:
  -h, --help   help for audit
```

> If the `[audit-id]` flag is not specified, the command shows a list of commands that have been executed. For example:

Reviewer: ```suggestion
If the `[audit-id]` argument is not specified, the command shows a list of commands that have been executed. For example:
```

{{< copyable "shell-regular" >}}

```bash
tiup dm audit
```

```
ID      Time                  Command
--      ----                  -------
4D5kQY  2020-08-13T05:38:19Z  tiup dm display test
4D5kNv  2020-08-13T05:36:13Z  tiup dm list
4D5kNr  2020-08-13T05:36:10Z  tiup dm deploy -p prod-cluster v2.0.0 ./examples/dm/minimal.yaml
```

> The first column is `audit-id`. To view the execution log of a certain command, pass the `audit-id` of a command as the flag as follows:

Reviewer: ```suggestion
The first column is `audit-id`. To view the execution log of a certain command, pass the `audit-id` argument as follows:
```

{{< copyable "shell-regular" >}}

```bash
tiup dm audit 4D5kQY
```

## Run commands on a host in the DM cluster

> To run command on a host in the TiDB cluster, use the `exec` command. The usage of the `exec` command is as follows:

Author: ```suggestion
To run commands on a host in the DM cluster, use the `exec` command. The usage of the `exec` command is as follows:
```

```bash
Usage:
  tiup dm exec <cluster-name> [flags]

Flags:
      --command string   the command run on cluster host (default "ls")
  -h, --help             help for exec
  -N, --node strings     Only exec on host with specified nodes
  -R, --role strings     Only exec on host with specified roles
      --sudo             use root permissions (default false)
```

For example, to execute `ls /tmp` on all DM nodes, run the following command:

{{< copyable "shell-regular" >}}

```bash
tiup dm exec prod-cluster --command='ls /tmp'
```

## dmctl

TiUP integrates the DM cluster controller `dmctl`.

Run the following command to use dmctl:

```bash
tiup dmctl [args]
```

Specify the version of dmctl:

```
tiup dmctl:v2.0.0 [args]
```

> For example, to add a source previously, the dmctl command is `dmctl --master-addr master1:8261 operate-source create /tmp/source1.yml`. After dmctl is integrated into TiUP, this command is:

Reviewer: ```suggestion
The previous dmctl command to add a source is `dmctl --master-addr master1:8261 operate-source create /tmp/source1.yml`. After dmctl is integrated into TiUP, the command is:
```

{{< copyable "shell-regular" >}}

```bash
tiup dmctl --master-addr master1:8261 operate-source create /tmp/source1.yml
```

## Use the system's native SSH client to connect to cluster

> All operations above performed on the cluster machine use the TiUP-embedded SSH client to connect to the cluster and execute commands. However, in some scenarios, you might also need to use the SSH client native to the control machine system to perform such cluster operations. For example:

Reviewer: ```suggestion
All operations above performed on the cluster machine use the SSH client embedded in TiUP to connect to the cluster and execute commands. However, in some scenarios, you might also need to use the SSH client native to the control machine system to perform such cluster operations. For example:
```

- To use a SSH plug-in for authentication
- To use a customized SSH client

Then you can use the `--native-ssh` command-line flag to enable the system-native command-line tool:

> - Deploy a cluster:  `tiup dm deploy <cluster-name> <version> <topo> --native-ssh`

Reviewer: ```suggestion
- Deploy a cluster: `tiup dm deploy <cluster-name> <version> <topo> --native-ssh`
```

- Start a cluster: `tiup dm start <cluster-name> --native-ssh`
- Upgrade a cluster: `tiup dm upgrade ... --native-ssh`

You can add `--native-ssh` in all cluster operation commands above to use the system's native SSH client.

To avoid adding such a flag in every command, you can use the `TIUP_NATIVE_SSH` system variable to specify whether to use the local SSH client:

```sh
export TIUP_NATIVE_SSH=true
# or
export TIUP_NATIVE_SSH=1
# or
export TIUP_NATIVE_SSH=enable
```

> If you specify this environmental variable and `--native-ssh` at the same time, `--native-ssh` prevails.

Reviewer: ```suggestion
If you specify this environment variable and `--native-ssh` at the same time, `--native-ssh` has higher priority.
```
